### PR TITLE
Add DiscoveryClient async methods and General Sync Cleanup

### PR DESCRIPTION
**Pull request by nathanwoctopusdeploy**
# Background

Add async methods for the DiscoveryClient
Mark sync methods/classes as Obsolete where an async version exists

# How to review this PR

Quality :heavy_check_mark:

# Pre-requisites

- [ ] I have read [How we use GitHub Issues](https://github.com/OctopusDeploy/Issues/blob/master/docs/CONTRIBUTING.internal.md) for help deciding when and where it's appropriate to make an issue.
- [ ] I have considered informing or consulting the right people, according to the [ownership map](https://whimsical.com/ownership-map-NzbiD4HJyvhC9jNJNfS6TG).
- [ ] I have considered appropriate testing for my change.

### CODE REVIEW

File: source/Halibut.Tests/Support/AssertAsync.cs [hunk 1]
```diff
@@ -0,0 +1,15 @@
+﻿using System;
+using System.Threading.Tasks;
+using FluentAssertions;
+using FluentAssertions.Specialized;
+
+namespace Halibut.Tests.Support
+{
+    public static class AssertAsync
```
Comment (nathanwoctopusdeploy) [inline source/Halibut.Tests/Support/AssertAsync.cs:8]: Copied from another PR authored by @LukeButters

Comment (nathanwoctopusdeploy) [reply source/Halibut.Tests/Support/AssertAsync.cs:8]: Added `because`

File: source/Halibut/HalibutRuntimeBuilder.cs [hunk 2]
```diff
@@ -79,7 +79,9 @@ public HalibutRuntime Build()
             var serviceFactory = this.serviceFactory ?? new NullServiceFactory();
             if (serverCertificate == null) throw new ArgumentException($"Set a server certificate with {nameof(WithServerCertificate)} before calling {nameof(Build)}", nameof(serverCertificate));
             var logFactory = this.logFactory ?? new LogFactory();
-            var queueFactory = this.queueFactory ?? new DefaultPendingRequestQueueFactory(logFactory);
+#pragma warning disable CS0612
+            var queueFactory = this.queueFactory ?? (asyncHalibutFeature.IsEnabled() ? new PendingRequestQueueFactoryAsync(logFactory) : new DefaultPendingRequestQueueFactory(logFactory));
```
Comment (nathanwoctopusdeploy) [inline source/Halibut/HalibutRuntimeBuilder.cs:83]: If we want an async HalibutRuntime and have not specified an `IPendingRequestQueue` then get the async one by default

**Review by sburmanoctopus — APPROVED** [review]
LGTM

**Pull request merged**